4-1 (так же и 3-1)

TanVD opened the pull request.

TanVD committed:
First functionality added

TanVD committed:
Problems with removing of bullet. Trying to fix

TanVD committed:
1. Hot fix for bullet problem. May be problem with memory.
2. Ground added
3. Method for placing on ground objects added
4. Bullets now explodes on ground
5. Few improvements

TanVD committed:
1. Added Multiplayer
2. Some classes were moved to directories
3. Need to remove methods for old way of multiplayer
4.  Need to refactor

TanVD committed:
1. Added menu
2.Refactoring made
3.A few improvements

TanVD committed:
1. Added documentation where needed;
2. Fixed some problems with adding and removing items from scene
3. Upgraded menu
4. Fixed radius of explosion visible accroding to collide algorithm
5. Another minor improvements
@

jzuken commented:
29 ворнингов это сурово. давайте сделаем, чтобы их вообще не было

jzuken commented on 2_year/1_term/4-1/Rules/GameRules.cpp:
```diff
@@ -0,0 +1,43 @@
+            {
+                if (itemCollides != item && image->collidesWithItem(itemCollides))
+                {
+                    if (dynamic_cast<AvatarImage *>(itemCollides) != nullptr)
```
откастить в переменную, конечно, не вариант. будем по два раза кастить

jzuken commented:
давайте сюда еще диаграмму классов хотя бы нарисуем

jzuken commented on 2_year/1_term/4-1/tankgamewindow.cpp:
```diff
@@ -0,0 +1,116 @@
+    ui->graphicsView->setScene(scene);
+    ui->graphicsView->installEventFilter(manager);
+
+    Avatar *tankFirst;
```
о, неинициализированные указатели

jzuken commented on 2_year/1_term/4-1/Controls/KeyControls/KeyControl.h:
```diff
@@ -0,0 +1,36 @@
+
+class KeyControl : public AvatarControl
+{
+    ///QString includes keys in specific oreder "L(eft)R(ight)U(p)D(own)F(aster)S(lower)F(ire)"
```
что, серьезно? а при создании я вижу
KeyControl *keyControl = new KeyControl("ADWSQE 1", this);
лучше уж никакой документации, чем неправильная

jzuken replied:
ну и getCommandFromKey() порядок двух последних другой

TanVD committed:
1. Get rid of warinigs (most of them, some a related to paint in which i don't need to use all parameters)
2. Repaired docs for KeyControls and added example
3. Added UML Diagram in png (vpp already was)
4. Updated UML Diagram
5. Minor improvements
